src/core/control/Control.cpp

```diff
@@ -1078,6 +1091,19 @@ void Control::makeGeometryTool(GeometryToolType tool) {
             geometryTool->notify();
             break;
         }
+        case COMPASS: {
+            auto compass = new Compass();
+            view->addOverlayView(std::make_unique<xoj::view::CompassView>(compass, view, zoom));
+            this->geometryTool = std::unique_ptr<GeometryTool>(compass);
+            this->geometryToolController = std::make_unique<CompassController>(view, compass);
+            std::unique_ptr<GeometryToolInputHandler> geometryToolInputHandler =
+                    std::make_unique<CompassInputHandler>(this->win->getXournal(), geometryToolController.get());
+            geometryToolInputHandler->registerToPool(compass->getHandlerPool());
+            xournal->input->setGeometryToolInputHandler(std::move(geometryToolInputHandler));
+            fireActionSelected(GROUP_GEOMETRY_TOOL, ACTION_COMPASS);
+            geometryTool->notify();
+            break;
```

Comment:
This shares a lot of code with the setsquare case. Could some of the common parts be extracted out?

Reply:
Also, I noticed that on `master` branch (as of `d3cd981a`), the setsquare is not initially repainted when toggled. I imagine that this issue would also affect the compass. Here is one possible fix:
```suggestion
            Range geomRange = geometryTool->getToolRange(/*transformed=*/true);
            if (!geomRange.empty()) {
                view->repaintRect(geomRange.getX(), geomRange.getY(), geomRange.getWidth(), geomRange.getHeight());
            }
            break;
```

Reply:
> This shares a lot of code with the setsquare case. Could some of the common parts be extracted out?

Probably it can be done using templates. Not sure we want that. I will make a suggestion.

Reply:
> Also, I noticed that on `master` branch (as of `d3cd981a`), the setsquare is not initially repainted when toggled. I imagine that this issue would also affect the compass. 

That should happen via `geometryTool->notify()`, which dispatches a `FLAG_DIRTY_REGION` request that repaints the area in `XojPageView::flagDirtyRegion` through `GeometryToolView::on(FlagDirtyRegionRequest, const Range& rg)`. As far as I can observe in current master, the setsquare is repainted immediately when toggled. Is it different on your system?

Reply:
> > This shares a lot of code with the setsquare case. Could some of the common parts be extracted out?
> 
> Probably it can be done using templates. Not sure we want that. I will make a suggestion.

Done!

Reply:
> As far as I can observe in current master, the setsquare is repainted immediately when toggled. Is it different on your system?

Yeah. On my NixOS 21.11 machine running GNOME Wayland, it doesn't appear until the page is repainted (e.g., by switching tools).

Reply:
> Probably it can be done using templates. Not sure we want that. I will make a suggestion.

The proposed function looks good to me as-is, but it does have the following drawback: it forces the view, controller, and input handler constructors to take a specific form. If we want to add another geometry tool in the future that requires more constructor parameters, this won't work anymore and we'll have to move the initialization out of the helper function.

Reply:
> > As far as I can observe in current master, the setsquare is repainted immediately when toggled. Is it different on your system?
> 
> Yeah. On my NixOS 21.11 machine running GNOME Wayland, it doesn't appear until the page is repainted (e.g., by switching tools).

I found the bug(s). Those are fixed in #4532. Therefore I'll wait for #4532 to get merged before proceeding with the present PR.

Reply:
#4532 is merged, which solves the repainting bug(s).